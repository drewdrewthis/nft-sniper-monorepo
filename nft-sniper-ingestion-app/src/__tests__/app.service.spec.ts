import { HttpModule } from '@nestjs/axios';
import { Test, TestingModule } from '@nestjs/testing';
import { ResevoirService } from '../apis/resevoir';
import { AppService } from '../app.service';
import { AlchemyService } from '../apis/alchemy/alchemy.service';
import { CacheModule } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { ResevoirModule } from '../apis/resevoir/resevoir.module';
import { HttpRecordingService } from '../http-recording/http-recording.service';

jest.setTimeout(15000);

/**
 * For the golden tests to work here, you'll need to import the API keys, or:
 * yarn dotenv -e .env yarn jest nft-sniper-ingestion-app/src/__tests__/app.service.spec.ts
 */

const tokens = [
  {
    contractAddress: '0x60E4d786628Fea6478F785A6d7e704777c86a7c6',
    tokenId: 9018,
  },
  {
    contractAddress: '0x60E4d786628Fea6478F785A6d7e704777c86a7c6',
    tokenId: 3927,
  },
  {
    contractAddress: '0xed5af388653567af2f388e6224dc7c4b3241c544',
    tokenId: 2743,
  },
  {
    contractAddress: '0x394E3d3044fC89fCDd966D3cb35Ac0B32B0Cda91',
    tokenId: 6781,
  },
];

describe('ResevoirService', () => {
  let service: AppService;
  let httpRecordingService: HttpRecordingService;

  beforeAll(() => {
    httpRecordingService = new HttpRecordingService(__filename);
  });

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [HttpModule, CacheModule.register(), ResevoirModule],
      providers: [AppService, ResevoirService, AlchemyService, ConfigService],
    }).compile();

    service = module.get<AppService>(AppService);
  });

  afterAll(() => {
    httpRecordingService.tearDown();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should be able to get nft demo data', async () => {
    const result = await service.getNftDemoData(tokens);
    expect(result).toMatchSnapshot();
  });
});
