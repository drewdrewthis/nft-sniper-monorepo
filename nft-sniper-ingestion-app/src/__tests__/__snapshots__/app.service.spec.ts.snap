// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResevoirService should be able to get nft demo data 1`] = `
[
  {
    "currentOwner": {
      "owners": [
        {
          "address": "0x828cebb8cc44cb9047f0f057f9603c7348553d84",
          "ownership": {
            "floorAskPrice": 11.7,
            "onSaleCount": "1",
            "tokenCount": "1",
            "topBidValue": 0.191,
            "totalBidValue": 0.191,
          },
        },
      ],
    },
    "highestBid": {
      "contract": "0x60e4d786628fea6478f785a6d7e704777c86a7c6",
      "createdAt": "2022-11-16T18:57:39.826Z",
      "expiration": 1668884243,
      "feeBps": 200,
      "feeBreakdown": [
        {
          "bps": 150,
          "kind": "marketplace",
          "recipient": "0x5924a28caaf1cc016617874a2f0c3710d881f3c1",
        },
        {
          "bps": 50,
          "kind": "royalty",
          "recipient": "0xaae7ac476b117bccafe2f05f582906be44bc8ff1",
        },
      ],
      "id": "0xf98ba2714ead81d2ed413a3250abdac144819a7bade177f9c5c4dcb201914587",
      "isDynamic": false,
      "isReservoir": null,
      "kind": "looks-rare",
      "maker": "0x828cebb8cc44cb9047f0f057f9603c7348553d84",
      "price": {
        "amount": {
          "decimal": 11.7,
          "native": 11.7,
          "raw": "11700000000000000000",
          "usd": 14247.11444,
        },
        "currency": {
          "contract": "0x0000000000000000000000000000000000000000",
          "decimals": 18,
          "name": "Ether",
          "symbol": "ETH",
        },
        "netAmount": {
          "decimal": 11.466,
          "native": 11.466,
          "raw": "11466000000000000000",
          "usd": 13962.17215,
        },
      },
      "quantityFilled": 0,
      "quantityRemaining": 1,
      "side": "sell",
      "source": {
        "domain": "looksrare.org",
        "icon": "https://raw.githubusercontent.com/reservoirprotocol/indexer/v5/src/models/sources/looksrare-logo.svg",
        "id": "0x5924a28caaf1cc016617874a2f0c3710d881f3c1",
        "name": "LooksRare",
        "url": "https://looksrare.org/collections/0x60e4d786628fea6478f785a6d7e704777c86a7c6/9018",
      },
      "status": "active",
      "taker": "0x0000000000000000000000000000000000000000",
      "tokenSetId": "token:0x60e4d786628fea6478f785a6d7e704777c86a7c6:9018",
      "tokenSetSchemaHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "updatedAt": "2022-11-16T18:57:39.826Z",
      "validFrom": 1668625048,
      "validUntil": 1668884243,
    },
    "lastSale": {
      "amount": "1",
      "batchIndex": 1,
      "block": 15984331,
      "fillSource": "looksrare.org",
      "from": "0xc05669e488f0942ff1c3e2f6921a7a1942dd6ca2",
      "id": "be81dec8d58965084b7e3736a8b821b568c88979e55c6edb09785febcddefc55",
      "logIndex": 191,
      "orderId": "0xd9b58bc42f6a46fb773f308b2fb670996b4ca01a3073e798e46e2ce0f94b185b",
      "orderKind": "looks-rare",
      "orderSide": "bid",
      "orderSource": "looksrare.org",
      "price": {
        "amount": {
          "decimal": 10.2,
          "native": 10.2,
          "raw": "10200000000000000000",
          "usd": 12759.53251,
        },
        "currency": {
          "contract": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
          "decimals": 18,
          "name": "Wrapped Ether",
          "symbol": "WETH",
        },
      },
      "saleId": "29f5f08f5f7bf839d8022687070dccc3155de7e3f3c6f083af60861851bce25d",
      "timestamp": 1668622847,
      "to": "0x828cebb8cc44cb9047f0f057f9603c7348553d84",
      "token": {
        "collection": {
          "id": null,
          "name": null,
        },
        "contract": "0x60e4d786628fea6478f785a6d7e704777c86a7c6",
        "image": null,
        "name": null,
        "tokenId": "9018",
      },
      "txHash": "0x215ae227621e57f77659d9dacd6339f574034f4882c03a761cd3e07ada2e7802",
      "washTradingScore": 0,
    },
    "lowestListing": {
      "contract": "0x60e4d786628fea6478f785a6d7e704777c86a7c6",
      "createdAt": "2022-11-16T18:57:39.826Z",
      "expiration": 1668884243,
      "feeBps": 200,
      "feeBreakdown": [
        {
          "bps": 150,
          "kind": "marketplace",
          "recipient": "0x5924a28caaf1cc016617874a2f0c3710d881f3c1",
        },
        {
          "bps": 50,
          "kind": "royalty",
          "recipient": "0xaae7ac476b117bccafe2f05f582906be44bc8ff1",
        },
      ],
      "id": "0xf98ba2714ead81d2ed413a3250abdac144819a7bade177f9c5c4dcb201914587",
      "isDynamic": false,
      "isReservoir": null,
      "kind": "looks-rare",
      "maker": "0x828cebb8cc44cb9047f0f057f9603c7348553d84",
      "price": {
        "amount": {
          "decimal": 11.7,
          "native": 11.7,
          "raw": "11700000000000000000",
          "usd": 14247.11444,
        },
        "currency": {
          "contract": "0x0000000000000000000000000000000000000000",
          "decimals": 18,
          "name": "Ether",
          "symbol": "ETH",
        },
        "netAmount": {
          "decimal": 11.466,
          "native": 11.466,
          "raw": "11466000000000000000",
          "usd": 13962.17215,
        },
      },
      "quantityFilled": 0,
      "quantityRemaining": 1,
      "side": "sell",
      "source": {
        "domain": "looksrare.org",
        "icon": "https://raw.githubusercontent.com/reservoirprotocol/indexer/v5/src/models/sources/looksrare-logo.svg",
        "id": "0x5924a28caaf1cc016617874a2f0c3710d881f3c1",
        "name": "LooksRare",
        "url": "https://looksrare.org/collections/0x60e4d786628fea6478f785a6d7e704777c86a7c6/9018",
      },
      "status": "active",
      "taker": "0x0000000000000000000000000000000000000000",
      "tokenSetId": "token:0x60e4d786628fea6478f785a6d7e704777c86a7c6:9018",
      "tokenSetSchemaHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "updatedAt": "2022-11-16T18:57:39.826Z",
      "validFrom": 1668625048,
      "validUntil": 1668884243,
    },
    "metadata": {
      "contract": {
        "address": "0x60e4d786628fea6478f785a6d7e704777c86a7c6",
      },
      "contractMetadata": {
        "name": "MutantApeYachtClub",
        "openSea": {
          "collectionName": "Mutant Ape Yacht Club",
          "description": "The MUTANT APE YACHT CLUB is a collection of up to 20,000 Mutant Apes that can only be created by exposing an existing Bored Ape to a vial of MUTANT SERUM or by minting a Mutant Ape in the public sale.",
          "floorPrice": 10.2,
          "imageUrl": "https://i.seadn.io/gae/lHexKRMpw-aoSyB1WdFBff5yfANLReFxHzt1DOj_sg7mS14yARpuvYcUtsyyx-Nkpk6WTcUPFoG53VnLJezYi8hAs0OxNZwlw6Y-dmI?w=500&auto=format",
          "lastIngestedAt": "2022-11-15T22:37:11.000Z",
          "safelistRequestStatus": "verified",
        },
        "symbol": "MAYC",
        "tokenType": "ERC721",
        "totalSupply": "19427",
      },
      "description": "",
      "id": {
        "tokenId": "9018",
        "tokenMetadata": {
          "tokenType": "ERC721",
        },
      },
      "media": [
        {
          "bytes": 189947,
          "format": "png",
          "gateway": "https://res.cloudinary.com/alchemyapi/image/upload/mainnet/b2c406c6d7593fe4d1902d19d30dae6f.png",
          "raw": "ipfs://QmZXdVH9Hc7GCQ615r9qyTwDSYwPgporHc16SCZqV6KMNa",
          "thumbnail": "https://res.cloudinary.com/alchemyapi/image/upload/w_256,h_256/mainnet/b2c406c6d7593fe4d1902d19d30dae6f.png",
        },
      ],
      "metadata": {
        "attributes": [
          {
            "trait_type": "Background",
            "value": "M1 Purple",
          },
          {
            "trait_type": "Fur",
            "value": "M1 Robot",
          },
          {
            "trait_type": "Eyes",
            "value": "M1 Sad",
          },
          {
            "trait_type": "Clothes",
            "value": "M1 Prison Jumpsuit",
          },
          {
            "trait_type": "Mouth",
            "value": "M1 Tongue Out",
          },
        ],
        "image": "ipfs://QmZXdVH9Hc7GCQ615r9qyTwDSYwPgporHc16SCZqV6KMNa",
      },
      "timeLastUpdated": "2022-11-16T18:20:51.547Z",
      "title": "",
      "tokenUri": {
        "gateway": "https://boredapeyachtclub.com/api/mutants/9018",
        "raw": "https://boredapeyachtclub.com/api/mutants/9018",
      },
    },
  },
  {
    "currentOwner": {
      "owners": [
        {
          "address": "0x165896732df25087eba4e736604559ceafdfc82f",
          "ownership": {
            "floorAskPrice": 11.99,
            "onSaleCount": "1",
            "tokenCount": "1",
            "topBidValue": 0.191,
            "totalBidValue": 0.191,
          },
        },
      ],
    },
    "highestBid": {
      "contract": "0x60e4d786628fea6478f785a6d7e704777c86a7c6",
      "createdAt": "2022-11-17T14:18:41.517Z",
      "expiration": 1671237167,
      "feeBps": 200,
      "feeBreakdown": [
        {
          "bps": 150,
          "kind": "marketplace",
          "recipient": "0x5924a28caaf1cc016617874a2f0c3710d881f3c1",
        },
        {
          "bps": 50,
          "kind": "royalty",
          "recipient": "0xaae7ac476b117bccafe2f05f582906be44bc8ff1",
        },
      ],
      "id": "0xc67f34389011da376ee520f37f503463761c2510b88d71b634eaa05b2012542e",
      "isDynamic": false,
      "isReservoir": null,
      "kind": "looks-rare",
      "maker": "0x165896732df25087eba4e736604559ceafdfc82f",
      "price": {
        "amount": {
          "decimal": 11.99,
          "native": 11.99,
          "raw": "11990000000000000000",
          "usd": 14600.24805,
        },
        "currency": {
          "contract": "0x0000000000000000000000000000000000000000",
          "decimals": 18,
          "name": "Ether",
          "symbol": "ETH",
        },
        "netAmount": {
          "decimal": 11.7502,
          "native": 11.7502,
          "raw": "11750200000000000000",
          "usd": 14308.24309,
        },
      },
      "quantityFilled": 0,
      "quantityRemaining": 1,
      "side": "sell",
      "source": {
        "domain": "looksrare.org",
        "icon": "https://raw.githubusercontent.com/reservoirprotocol/indexer/v5/src/models/sources/looksrare-logo.svg",
        "id": "0x5924a28caaf1cc016617874a2f0c3710d881f3c1",
        "name": "LooksRare",
        "url": "https://looksrare.org/collections/0x60e4d786628fea6478f785a6d7e704777c86a7c6/3927",
      },
      "status": "active",
      "taker": "0x0000000000000000000000000000000000000000",
      "tokenSetId": "token:0x60e4d786628fea6478f785a6d7e704777c86a7c6:3927",
      "tokenSetSchemaHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "updatedAt": "2022-11-17T14:18:41.517Z",
      "validFrom": 1668694687,
      "validUntil": 1671237167,
    },
    "lastSale": {
      "amount": "1",
      "batchIndex": 1,
      "block": 15096851,
      "fillSource": "looksrare.org",
      "from": "0x4a39e17ad4b045216d3702b2748e2c26f597fd84",
      "id": "decb97716f04efec2e8381f3826d453a34c5962e05e198f8df78837d898014cb",
      "logIndex": 56,
      "orderId": "0x84f65adaa01ecdb8183ad525505056622bbdab625336dd31274762d8232e56f3",
      "orderKind": "looks-rare",
      "orderSide": "ask",
      "orderSource": "looksrare.org",
      "price": {
        "amount": {
          "decimal": 18,
          "native": 18,
          "raw": "18000000000000000000",
          "usd": 21397.06157,
        },
        "currency": {
          "contract": "0x0000000000000000000000000000000000000000",
          "decimals": 18,
          "name": "Ether",
          "symbol": "ETH",
        },
      },
      "saleId": "d49259c22437d2a29085a30625d19f430e300faee764ef1c9fef3e4a9cba666d",
      "timestamp": 1657215937,
      "to": "0x165896732df25087eba4e736604559ceafdfc82f",
      "token": {
        "collection": {
          "id": null,
          "name": null,
        },
        "contract": "0x60e4d786628fea6478f785a6d7e704777c86a7c6",
        "image": null,
        "name": null,
        "tokenId": "3927",
      },
      "txHash": "0x3337067167986befd021710b1cfe705141dceeb8e1edfc6a09c91591c1f73d48",
      "washTradingScore": 0,
    },
    "lowestListing": {
      "contract": "0x60e4d786628fea6478f785a6d7e704777c86a7c6",
      "createdAt": "2022-11-17T00:33:16.259Z",
      "expiration": 1671237167,
      "feeBps": 200,
      "feeBreakdown": [
        {
          "bps": 150,
          "kind": "marketplace",
          "recipient": "0x5924a28caaf1cc016617874a2f0c3710d881f3c1",
        },
        {
          "bps": 50,
          "kind": "royalty",
          "recipient": "0xaae7ac476b117bccafe2f05f582906be44bc8ff1",
        },
      ],
      "id": "0x93a05ddafd403917a08180eef9af66d7fabdf35b3da114d7ee1482fa4acf3fab",
      "isDynamic": false,
      "isReservoir": null,
      "kind": "looks-rare",
      "maker": "0x165896732df25087eba4e736604559ceafdfc82f",
      "price": {
        "amount": {
          "decimal": 12.4,
          "native": 12.4,
          "raw": "12400000000000000000",
          "usd": 15099.5059,
        },
        "currency": {
          "contract": "0x0000000000000000000000000000000000000000",
          "decimals": 18,
          "name": "Ether",
          "symbol": "ETH",
        },
        "netAmount": {
          "decimal": 12.152,
          "native": 12.152,
          "raw": "12152000000000000000",
          "usd": 14797.51578,
        },
      },
      "quantityFilled": 0,
      "quantityRemaining": 1,
      "side": "sell",
      "source": {
        "domain": "looksrare.org",
        "icon": "https://raw.githubusercontent.com/reservoirprotocol/indexer/v5/src/models/sources/looksrare-logo.svg",
        "id": "0x5924a28caaf1cc016617874a2f0c3710d881f3c1",
        "name": "LooksRare",
        "url": "https://looksrare.org/collections/0x60e4d786628fea6478f785a6d7e704777c86a7c6/3927",
      },
      "status": "active",
      "taker": "0x0000000000000000000000000000000000000000",
      "tokenSetId": "token:0x60e4d786628fea6478f785a6d7e704777c86a7c6:3927",
      "tokenSetSchemaHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "updatedAt": "2022-11-17T00:33:16.259Z",
      "validFrom": 1668645176,
      "validUntil": 1671237167,
    },
    "metadata": {
      "contract": {
        "address": "0x60e4d786628fea6478f785a6d7e704777c86a7c6",
      },
      "contractMetadata": {
        "name": "MutantApeYachtClub",
        "openSea": {
          "collectionName": "Mutant Ape Yacht Club",
          "description": "The MUTANT APE YACHT CLUB is a collection of up to 20,000 Mutant Apes that can only be created by exposing an existing Bored Ape to a vial of MUTANT SERUM or by minting a Mutant Ape in the public sale.",
          "floorPrice": 10.2,
          "imageUrl": "https://i.seadn.io/gae/lHexKRMpw-aoSyB1WdFBff5yfANLReFxHzt1DOj_sg7mS14yARpuvYcUtsyyx-Nkpk6WTcUPFoG53VnLJezYi8hAs0OxNZwlw6Y-dmI?w=500&auto=format",
          "lastIngestedAt": "2022-11-15T22:37:11.000Z",
          "safelistRequestStatus": "verified",
        },
        "symbol": "MAYC",
        "tokenType": "ERC721",
        "totalSupply": "19427",
      },
      "description": "",
      "id": {
        "tokenId": "3927",
        "tokenMetadata": {
          "tokenType": "ERC721",
        },
      },
      "media": [
        {
          "gateway": "https://ipfs.io/ipfs/QmR1EUAXrH62WZEuBSGa3so8NpdYv6i2fdDfBktGcZL2f3",
          "raw": "ipfs://QmR1EUAXrH62WZEuBSGa3so8NpdYv6i2fdDfBktGcZL2f3",
        },
      ],
      "metadata": {
        "attributes": [
          {
            "trait_type": "Background",
            "value": "M1 Blue",
          },
          {
            "trait_type": "Fur",
            "value": "M1 Dark Brown",
          },
          {
            "trait_type": "Eyes",
            "value": "M1 Angry",
          },
          {
            "trait_type": "Clothes",
            "value": "M1 Tanktop",
          },
          {
            "trait_type": "Hat",
            "value": "M1 Halo",
          },
          {
            "trait_type": "Mouth",
            "value": "M1 Phoneme L",
          },
        ],
        "image": "ipfs://QmR1EUAXrH62WZEuBSGa3so8NpdYv6i2fdDfBktGcZL2f3",
      },
      "timeLastUpdated": "2022-11-17T16:14:38.699Z",
      "title": "",
      "tokenUri": {
        "gateway": "https://boredapeyachtclub.com/api/mutants/3927",
        "raw": "https://boredapeyachtclub.com/api/mutants/3927",
      },
    },
  },
  {
    "currentOwner": {
      "owners": [
        {
          "address": "0x8599254f994d008f77a2144e049ee27940e9045f",
          "ownership": {
            "floorAskPrice": 12,
            "onSaleCount": "1",
            "tokenCount": "1",
            "topBidValue": 0.186,
            "totalBidValue": 0.186,
          },
        },
      ],
    },
    "highestBid": {
      "contract": "0xed5af388653567af2f388e6224dc7c4b3241c544",
      "createdAt": "2022-11-17T05:00:26.383Z",
      "expiration": 1668920260,
      "feeBps": 750,
      "feeBreakdown": [
        {
          "bps": 250,
          "kind": "marketplace",
          "recipient": "0x0000a26b00c1f0df003000390027140000faa719",
        },
        {
          "bps": 500,
          "kind": "royalty",
          "recipient": "0xb4d24dacbdffa1bbf9a624044484b3feeb7fdf74",
        },
      ],
      "id": "0x239af5f8e335b2652a106bbecb141abcee5e195de625261c143204b645381fcb",
      "isDynamic": false,
      "isReservoir": null,
      "kind": "seaport",
      "maker": "0x8599254f994d008f77a2144e049ee27940e9045f",
      "price": {
        "amount": {
          "decimal": 12,
          "native": 12,
          "raw": "12000000000000000000",
          "usd": 14612.42507,
        },
        "currency": {
          "contract": "0x0000000000000000000000000000000000000000",
          "decimals": 18,
          "name": "Ether",
          "symbol": "ETH",
        },
        "netAmount": {
          "decimal": 11.1,
          "native": 11.1,
          "raw": "11100000000000000000",
          "usd": 13516.49319,
        },
      },
      "quantityFilled": 0,
      "quantityRemaining": 1,
      "side": "sell",
      "source": {
        "domain": "opensea.io",
        "icon": "https://raw.githubusercontent.com/reservoirprotocol/indexer/v5/src/models/sources/opensea-logo.svg",
        "id": "0x5b3256965e7c3cf26e11fcaf296dfc8807c01073",
        "name": "OpenSea",
        "url": "https://opensea.io/assets/0xed5af388653567af2f388e6224dc7c4b3241c544/2743",
      },
      "status": "active",
      "taker": "0x0000000000000000000000000000000000000000",
      "tokenSetId": "token:0xed5af388653567af2f388e6224dc7c4b3241c544:2743",
      "tokenSetSchemaHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "updatedAt": "2022-11-17T05:00:53.391Z",
      "validFrom": 1668661060,
      "validUntil": 1668920260,
    },
    "lastSale": {
      "amount": "1",
      "batchIndex": 1,
      "block": 14735326,
      "fillSource": "opensea.io",
      "from": "0xb20c8e35497977e55469d3d20a8358a26d7857f8",
      "id": "4cf50858a0b196fbcda64205fdac10d8931087325bff25acd45830798f4cd597",
      "logIndex": 469,
      "orderId": "0x80c12a2f563fb3a847e9347eba692f07111b2bb9015939a6e8e2574fd43fcdde",
      "orderKind": "wyvern-v2.3",
      "orderSide": "ask",
      "orderSource": "opensea.io",
      "price": {
        "amount": {
          "decimal": 22,
          "native": 22,
          "raw": "22000000000000000000",
          "usd": 58107.04035,
        },
        "currency": {
          "contract": "0x0000000000000000000000000000000000000000",
          "decimals": 18,
          "name": "Ether",
          "symbol": "ETH",
        },
      },
      "saleId": "f999c07ef48a9b877e3eee60d09f003a2bcf360137ac99eb81f84f166bcdbc69",
      "timestamp": 1652003079,
      "to": "0x8599254f994d008f77a2144e049ee27940e9045f",
      "token": {
        "collection": {
          "id": null,
          "name": null,
        },
        "contract": "0xed5af388653567af2f388e6224dc7c4b3241c544",
        "image": null,
        "name": null,
        "tokenId": "2743",
      },
      "txHash": "0x6c150063091ad10b5e3dcf07333ba56f98e95fe0379ef063feea43e9d655ed3f",
      "washTradingScore": 0,
    },
    "lowestListing": {
      "contract": "0xed5af388653567af2f388e6224dc7c4b3241c544",
      "createdAt": "2022-11-17T05:00:26.383Z",
      "expiration": 1668920260,
      "feeBps": 750,
      "feeBreakdown": [
        {
          "bps": 250,
          "kind": "marketplace",
          "recipient": "0x0000a26b00c1f0df003000390027140000faa719",
        },
        {
          "bps": 500,
          "kind": "royalty",
          "recipient": "0xb4d24dacbdffa1bbf9a624044484b3feeb7fdf74",
        },
      ],
      "id": "0x239af5f8e335b2652a106bbecb141abcee5e195de625261c143204b645381fcb",
      "isDynamic": false,
      "isReservoir": null,
      "kind": "seaport",
      "maker": "0x8599254f994d008f77a2144e049ee27940e9045f",
      "price": {
        "amount": {
          "decimal": 12,
          "native": 12,
          "raw": "12000000000000000000",
          "usd": 14612.42507,
        },
        "currency": {
          "contract": "0x0000000000000000000000000000000000000000",
          "decimals": 18,
          "name": "Ether",
          "symbol": "ETH",
        },
        "netAmount": {
          "decimal": 11.1,
          "native": 11.1,
          "raw": "11100000000000000000",
          "usd": 13516.49319,
        },
      },
      "quantityFilled": 0,
      "quantityRemaining": 1,
      "side": "sell",
      "source": {
        "domain": "opensea.io",
        "icon": "https://raw.githubusercontent.com/reservoirprotocol/indexer/v5/src/models/sources/opensea-logo.svg",
        "id": "0x5b3256965e7c3cf26e11fcaf296dfc8807c01073",
        "name": "OpenSea",
        "url": "https://opensea.io/assets/0xed5af388653567af2f388e6224dc7c4b3241c544/2743",
      },
      "status": "active",
      "taker": "0x0000000000000000000000000000000000000000",
      "tokenSetId": "token:0xed5af388653567af2f388e6224dc7c4b3241c544:2743",
      "tokenSetSchemaHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "updatedAt": "2022-11-17T05:00:53.391Z",
      "validFrom": 1668661060,
      "validUntil": 1668920260,
    },
    "metadata": {
      "contract": {
        "address": "0xed5af388653567af2f388e6224dc7c4b3241c544",
      },
      "contractMetadata": {
        "name": "Azuki",
        "openSea": {
          "collectionName": "Azuki",
          "description": "Take the red bean to join the garden. View the collection at [azuki.com/gallery](https://azuki.com/gallery).

Azuki starts with a collection of 10,000 avatars that give you membership access to The Garden: a corner of the internet where artists, builders, and web3 enthusiasts meet to create a decentralized future. Azuki holders receive access to exclusive drops, experiences, and more. Visit [azuki.com](https://azuki.com) for more details.

We rise together. We build together. We grow together.",
          "discordUrl": "https://discord.gg/azuki",
          "externalUrl": "http://www.azuki.com",
          "floorPrice": 9.94,
          "imageUrl": "https://i.seadn.io/gae/H8jOCJuQokNqGBpkBN5wk1oZwO7LM8bNnrHCaekV2nKjnCqw6UB5oaH8XyNeBDj6bA_n1mjejzhFQUP3O1NfjFLHr3FOaeHcTOOT?w=500&auto=format",
          "lastIngestedAt": "2022-11-17T10:54:39.000Z",
          "safelistRequestStatus": "verified",
          "twitterUsername": "azukiofficial",
        },
        "symbol": "AZUKI",
        "tokenType": "ERC721",
        "totalSupply": "10000",
      },
      "description": "",
      "id": {
        "tokenId": "2743",
        "tokenMetadata": {
          "tokenType": "ERC721",
        },
      },
      "media": [
        {
          "gateway": "https://ipfs.io/ipfs/QmYDvPAXtiJg7s8JdRBSLWdgSphQdac8j1YuQNNxcGE1hg/2743.png",
          "raw": "https://ikzttp.mypinata.cloud/ipfs/QmYDvPAXtiJg7s8JdRBSLWdgSphQdac8j1YuQNNxcGE1hg/2743.png",
        },
      ],
      "metadata": {
        "attributes": [
          {
            "trait_type": "Type",
            "value": "Human",
          },
          {
            "trait_type": "Hair",
            "value": "Buzzcut",
          },
          {
            "trait_type": "Headgear",
            "value": "Straw Hat",
          },
          {
            "trait_type": "Clothing",
            "value": "Cloud Poncho",
          },
          {
            "trait_type": "Eyes",
            "value": "Focused",
          },
          {
            "trait_type": "Mouth",
            "value": "Toothpick",
          },
          {
            "trait_type": "Offhand",
            "value": "Bean Juice",
          },
          {
            "trait_type": "Background",
            "value": "Off White A",
          },
        ],
        "image": "https://ikzttp.mypinata.cloud/ipfs/QmYDvPAXtiJg7s8JdRBSLWdgSphQdac8j1YuQNNxcGE1hg/2743.png",
        "name": "Azuki #2743",
      },
      "timeLastUpdated": "2022-11-12T09:06:14.012Z",
      "title": "Azuki #2743",
      "tokenUri": {
        "gateway": "https://ipfs.io/ipfs/QmQFkLSQysj94s5GvTHPyzTxrawwtjgiiYS2TBLgrvw8CW/2743",
        "raw": "https://ikzttp.mypinata.cloud/ipfs/QmQFkLSQysj94s5GvTHPyzTxrawwtjgiiYS2TBLgrvw8CW/2743",
      },
    },
  },
  {
    "currentOwner": {
      "owners": [
        {
          "address": "0xe31bb1a9a271b754cbadcd5e993c883d8e4cf9af",
          "ownership": {
            "floorAskPrice": null,
            "onSaleCount": "0",
            "tokenCount": "1",
            "topBidValue": null,
            "totalBidValue": null,
          },
        },
      ],
    },
    "highestBid": undefined,
    "lastSale": {
      "amount": "1",
      "batchIndex": 1,
      "block": 15675893,
      "fillSource": "opensea.io",
      "from": "0xa1c0c0ff6eaa63e7b0c7bf441fc146afdb08fc7b",
      "id": "a99ef93fbe2da5ed84f255d96b94a28b054dfbb36e6b155fbd3290e4231121a5",
      "logIndex": 273,
      "orderId": "0x9f05f13ba7be414d52fd0652c18aceb08a86af18f3d5e62545afc51246f5b928",
      "orderKind": "seaport",
      "orderSide": "bid",
      "orderSource": "opensea.io",
      "price": {
        "amount": {
          "decimal": 5.3,
          "native": 5.3,
          "raw": "5300000000000000000",
          "usd": 7014.2535,
        },
        "currency": {
          "contract": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
          "decimals": 18,
          "name": "Wrapped Ether",
          "symbol": "WETH",
        },
      },
      "saleId": "19034aaa3cb554b941b79af25a40fdf16051fe33caed08d61b428fc1a294e118",
      "timestamp": 1664900099,
      "to": "0xe31bb1a9a271b754cbadcd5e993c883d8e4cf9af",
      "token": {
        "collection": {
          "id": null,
          "name": null,
        },
        "contract": "0x394e3d3044fc89fcdd966d3cb35ac0b32b0cda91",
        "image": null,
        "name": null,
        "tokenId": "6781",
      },
      "txHash": "0xa4823ab85eb06c6d6fa27e809a184249563719a63299fccbccda7dd61b116533",
      "washTradingScore": 0,
    },
    "lowestListing": undefined,
    "metadata": {
      "contract": {
        "address": "0x394e3d3044fc89fcdd966d3cb35ac0b32b0cda91",
      },
      "contractMetadata": {
        "name": "RENGA",
        "openSea": {
          "collectionName": "RENGA",
          "description": "A handcrafted collection of 10,000 characters developed by artist DirtyRobot. Each with their own identity to be discovered within the wider stories of RENGA. In its purest form, RENGA is the art of storytelling.",
          "discordUrl": "https://discord.gg/renga",
          "externalUrl": "https://renga.app/",
          "floorPrice": 0.958,
          "imageUrl": "https://i.seadn.io/gcs/files/f3b11e36be14a5d31c75b19d03996fed.gif?w=500&auto=format",
          "lastIngestedAt": "2022-11-15T11:55:26.000Z",
          "safelistRequestStatus": "verified",
          "twitterUsername": "RENGA_inc",
        },
        "symbol": "RENGA",
        "tokenType": "ERC721",
      },
      "description": "A handcrafted collection of 10,000 characters developed by artist DirtyRobot. Each with their own identity to be discovered within the wider stories of RENGA. In its purest form, RENGA is the art of storytelling. Visit our [website](https://renga.app/) for more details.",
      "id": {
        "tokenId": "6781",
        "tokenMetadata": {
          "tokenType": "ERC721",
        },
      },
      "media": [
        {
          "gateway": "https://nfts.renga.app/nfts/public/images/6781.jpeg",
          "raw": "https://nfts.renga.app/nfts/public/images/6781.jpeg",
        },
      ],
      "metadata": {
        "attributes": [
          {
            "trait_type": "Archetype",
            "value": "Samurai",
          },
          {
            "trait_type": "Background",
            "value": "Black",
          },
          {
            "trait_type": "Head",
            "value": "Marble",
          },
          {
            "trait_type": "Armour",
            "value": "Iron Armour Purple",
          },
          {
            "trait_type": "Nose",
            "value": "Ordinary",
          },
          {
            "trait_type": "Hair",
            "value": "Black",
          },
          {
            "trait_type": "Mouth",
            "value": "Cocky Grin",
          },
          {
            "trait_type": "Eyes",
            "value": "Bold",
          },
          {
            "trait_type": "Hands",
            "value": "Dragon Katana Black",
          },
        ],
        "description": "A handcrafted collection of 10,000 characters developed by artist DirtyRobot. Each with their own identity to be discovered within the wider stories of RENGA. In its purest form, RENGA is the art of storytelling. Visit our [website](https://renga.app/) for more details.",
        "image": "https://nfts.renga.app/nfts/public/images/6781.jpeg",
        "name": "RENGA #6781",
      },
      "timeLastUpdated": "2022-11-17T19:38:31.378Z",
      "title": "RENGA #6781",
      "tokenUri": {
        "gateway": "https://nfts.renga.app/nfts/public/erc-721/6781",
        "raw": "https://nfts.renga.app/nfts/public/erc-721/6781",
      },
    },
  },
]
`;
